core/src/main/java/org/apache/iceberg/LocationProviders.java

```diff
@@ -141,4 +133,18 @@ private static String stripTrailingSlash(String path) {
     }
     return result;
   }
+
+  public static String dataLocation(Map<String, String> properties, String tableLocation) {
+    String dataLocation = properties.get(TableProperties.WRITE_DATA_LOCATION);
+    if (dataLocation == null) {
+      dataLocation = properties.get(TableProperties.OBJECT_STORE_PATH);
```

Comment:
This is only correct if the table is using object storage. I think you need to update this logic to select whether to return OBJECT_STORE_PATH or WRITE_FOLDER_STORAGE_LOCATION depending on the location provider selected. Both should fall back to the table location. And the new property should take precedence.

Reply:
Agreed. Made the change. This is going to be a little different from #2965. Modified the test case `testObjectStorageLocationProviderPathResolution` and added `testDefaultStorageLocationProviderPathResolution`.